File: js/Advection.js
"use strict";

var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();

Object.defineProperty(exports, "__esModule", { value: true });

// import { BufferGeometry, BufferAttribute, RawShaderMaterial, LineSegments } from "three";
var face_vert_raw_1 = require("../shaders/face.vert?raw");
var line_vert_raw_1 = require("../shaders/line.vert?raw");
var advection_frag_raw_1 = require("../shaders/advection.frag?raw");
import ShaderPass_1 from "./ShaderPass.js";

var BufferGeometry = THREE.BufferGeometry
var BufferAttribute = THREE.BufferAttribute
var RawShaderMaterial = THREE.RawShaderMaterial
var LineSegments = THREE.LineSegments
var Advection = /** @class */ (function (_super) {
    __extends(Advection, _super);
    function Advection(simProps) {
        var _this = _super.call(this, {
            material: {
                vertexShader: face_vert_raw_1,
                fragmentShader: advection_frag_raw_1,
                uniforms: {
                    boundarySpace: {
                        value: simProps.cellScale,
                    },
                    px: {
                        value: simProps.cellScale,
                    },
                    fboSize: {
                        value: simProps.fboSize,
                    },
                    velocity: {
                        value: simProps.src.texture,
                    },
                    dt: {
                        value: simProps.dt,
                    },
                    isBFECC: {
                        value: true,
                    },
                },
            },
            output: simProps.dst,
        }) || this;
        _this.init();
        return _this;
    }
    Advection.prototype.init = function () {
        _super.prototype.init.call(this);
        this.createBoundary();
    };
    Advection.prototype.createBoundary = function () {
        var _a;
        var boundaryG = new BufferGeometry();
        var vertices_boundary = new Float32Array([
            // left
            -1, -1, 0, -1, 1, 0,
            // top
            -1, 1, 0, 1, 1, 0,
            // right
            1, 1, 0, 1, -1, 0,
            // bottom
            1, -1, 0, -1, -1, 0,
        ]);
        boundaryG.setAttribute("position", new BufferAttribute(vertices_boundary, 3));
        var boundaryM = new RawShaderMaterial({
            vertexShader: line_vert_raw_1,
            fragmentShader: advection_frag_raw_1,
            uniforms: this.uniforms,
        });
        this.line = new LineSegments(boundaryG, boundaryM);
        (_a = this.scene) === null || _a === void 0 ? void 0 : _a.add(this.line);
    };
    Advection.prototype.updateAdvection = function (_a) {
        var dt = _a.dt, isBounce = _a.isBounce, BFECC = _a.BFECC;
        if (this.uniforms)
            this.uniforms.dt.value = dt;
        this.line.visible = isBounce;
        this.uniforms.isBFECC.value = BFECC;
        _super.prototype.update.call(this);
    };
    return Advection;
}(ShaderPass_1));
const _default = Advection;
export { _default as default };

File: src/Advection.lua
local get = require("utils.shaders")
local face_vert = get("face.vert")
local line_vert = get("line.vert")
local advection_frag = get("advection.frag")

local ShaderPass = require("src.ShaderPass")
local new = require("utils.new")
local Object = require("utils.convertToJSObject")

local THREE = js.global.THREE

local function Float32Array(array)
    local a = new(js.global.Float32Array)

    for i, v in ipairs(array) do
        a[i-1] = v
    end

    return a
end

return function(simulationProperties)
    local self = ShaderPass(Object({
        material = {
            vertexShader = face_vert,
            fragmentShader = advection_frag,
            uniforms = {
                boundarySpace = {
                    value = simulationProperties.cellScale
                },
                px = {
                    value = simulationProperties.cellScale
                },
                fboSize = {
                    value = simulationProperties.fboSize
                },
                velocity = {
                    value = simulationProperties.src.texture
                },
                dt = {
                    value = simulationProperties.dt
                },
                isBFECC = {
                    value = true
                }
            },
        },
        output = simulationProperties.dst
    }))

    self.init()

    local boundaryG = new(THREE.BufferGeometry)
    local vertices_boundary = Float32Array({
        -- left
        -1, -1, 0,
        -1, 1, 0,

        -- top
        -1, 1, 0,
        1, 1, 0,

        -- right
        1, 1, 0,
        1, -1, 0,

        -- bottom
        1, -1, 0,
        -1, -1, 0
    })

    boundaryG:setAttribute("position", new(THREE.BufferAttribute, vertices_boundary, 3))

    local boundaryM = new(THREE.RawShaderMaterial, Object({
        vertexShader = line_vert,
        fragmentShader = advection_frag,
        uniforms = self.uniforms
    }))

    self.line = new(THREE.LineSegments, boundaryG, boundaryM)
    self.scene:add(self.line)

    self.update = function(dt, isBounce, BFECC)
        self.uniforms.dt.value = dt
        self.line.visible = isBounce
        self.uniforms.isBFECC.value = BFECC

        self._update()
    end

    print("Advection.lua initialized")
    return self
end